cmd/util/ledger/migrations/filter_unreferenced_slabs_migration.go

```diff
@@ -0,0 +1,158 @@
+		}
+
+		if !registerID.IsSlabIndex() {
+			continue
+		}
+
+		storageID := StorageIDFromRegisterID(registerID)
+
+		// Retrieve the slab.
+		_, _, err = storage.Retrieve(storageID)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve slab %s: %w", storageID, err)
+		}
+	}
+
+	// Load storage map.
+	for _, domain := range domains {
+		_ = storage.GetStorageMap(address, domain, false)
+	}
+
+	return storage.CheckHealth()
+}
+
+type FilterUnreferencedSlabsMigration struct {
+	log zerolog.Logger
+}
+
+var _ AccountBasedMigration = &FilterUnreferencedSlabsMigration{}
+
+func (m *FilterUnreferencedSlabsMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	m.log = log.
+		With().
+		Str("migration", "filter-unreferenced-slabs").
+		Logger()
+
+	return nil
+}
+
+func (m *FilterUnreferencedSlabsMigration) MigrateAccount(
+	_ context.Context,
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+
+	checkPayloadsOwnership(oldPayloads, address, m.log)
+
+	migrationRuntime, err := NewMigratorRuntime(
+		address,
+		oldPayloads,
+		util.RuntimeInterfaceConfig{},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create migrator runtime: %w", err)
+	}
+
+	storage := migrationRuntime.Storage
+
+	newPayloads := oldPayloads
+
+	err = CheckAccountStorageHealth(address, oldPayloads, storage)
+	if err != nil {
+
+		// The storage health check failed.
+		// This can happen if there are unreferenced root slabs.
+		// In this case, we filter out the unreferenced root slabs from the payloads.
+
+		var unreferencedRootSlabsErr runtime.UnreferencedRootSlabsError
+		if !errors.As(err, &unreferencedRootSlabsErr) {
+			return nil, fmt.Errorf("storage health check failed: %w", err)
+		}
+
+		m.log.Warn().
+			Err(err).
+			Str("account", address.Hex()).
+			Msg("filtering unreferenced root slabs")
+
+		// Create a set of unreferenced root slabs.
+
+		unreferencedRootSlabIDs := map[atree.StorageID]struct{}{}
+		for _, storageID := range unreferencedRootSlabsErr.UnreferencedRootSlabIDs {
+			unreferencedRootSlabIDs[storageID] = struct{}{}
+		}
+
+		// Filter out unreferenced root slabs.
+
+		newCount := len(oldPayloads) - len(unreferencedRootSlabIDs)
+		newPayloads = make([]*ledger.Payload, 0, newCount)
+
+		filteredPayloads := make([]*ledger.Payload, 0, len(unreferencedRootSlabIDs))
+
+		for _, payload := range oldPayloads {
+			registerID, _, err := convert.PayloadToRegister(payload)
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert payload to register: %w", err)
+			}
+
+			// Filter unreferenced slabs.
+			if registerID.IsSlabIndex() {
+				storageID := StorageIDFromRegisterID(registerID)
+				if _, ok := unreferencedRootSlabIDs[storageID]; ok {
+					filteredPayloads = append(filteredPayloads, payload)
+					continue
+				}
+			}
+
+			newPayloads = append(newPayloads, payload)
+		}
+
+		// TODO: write filtered payloads to a separate file
```

Comment:
@fxamacker @janezpodhostnik What would be a good way to write these filtered payloads? Should we create one payloads file per account? One file with all filtered payloads of all account?

Reply:
The report writer can be used.

I would create one file for all accounts with the report writer. The account address can be deduced from the payload key.

Reply:
@janezpodhostnik As far as I understand, the report writer generates JSON? Do you mean that we should  serialize the payloads to JSON (hex-encoded)?